interface.go

```diff
@@ -0,0 +1,76 @@
+	//
+	// See: Publish.
+	PublishLocal(ctx context.Context, adv schema.Advertisement) (cid.Cid, error)
+
+	// Publish appends adv to the locally stored advertisement chain, and publishes the new advertisement onto gossip pubsub.
+	// The CID returned represents the ID of the advertisement appended to the chain.
+	Publish(ctx context.Context, adv schema.Advertisement) (cid.Cid, error)
+
+	// Push pushes an update for a single entry.
+	// This can be used to perform updates for a small number of CIDs
+	// When a full advertisement is not worth it (web3.storage case).
+	// Indexer may only accept pushes from authenticated providers.
+	// TODO: determine if this function is safe to remove; may be used for testing purposes in storetheindex?
+	Push(ctx context.Context, indexer peer.ID, h mh.Multihash, metadata []byte) error
+
+	// RegisterCallback registers the callback used by the provider to look up a list of multihashes by LookupKey.
+	// Only a single callback is supported; repeated calls to this function will replace the previous callback.
+	RegisterCallback(cb Callback)
+
+	// NotifyPut sginals to the provider that the list of multihashes looked up by the given key are available.
+	// The given key is then used to look up the list of multihashes via Callback.
+	// An advertisement is then generated, appended to the chain of advertisements and published onto the gossip pubsub channel.
+	// Therefore, a Callback must be registered prior to using this function.
+	// The metadata may be nil and is optional.
+	//
+	// This function returns the ID of the advertisement published.
+	NotifyPut(ctx context.Context, key LookupKey, metadata []byte) (cid.Cid, error)
+
+	// NotifyRemove sginals to the provider that the multihashes that corresponded to the given key are no longer available.
+	// An advertisement is then generated, appended to the chain of advertisements and published onto the gossip pubsub channel.
+	// The given key must have previously been put via NotifyPut.
+	//
+	// This function returns the ID of the advertisement published.
+	NotifyRemove(ctx context.Context, key LookupKey) (cid.Cid, error)
+
+	// GetAdv gets the advertisement that corresponds to the given id.
+	GetAdv(ctx context.Context, id cid.Cid) (schema.Advertisement, error)
+
+	// GetLatestAdv gets the latest advertisement on this provider's advertisement chain and the CID to which it corresponds.
+	GetLatestAdv(ctx context.Context) (cid.Cid, schema.Advertisement, error)
+
+	// Close closes this provider, discarding all resources occupied by it.
+	// After calling this function the provider is no longer available for use.
+	Close(ctx context.Context) error
+}
+
+// Callback is used by provider to look up a list of multihashes associated to a key.
+// The callback must produce the same list of multihashes for the same key.
+// See: Interface.NotifyPut, Interface.NotifyRemove
+// TODO: update docs once the iterator return type refactor is done.
```

Comment:
To avoid extra noise here and to explicitly document the change in a commit of its own I will work on this in a separate PR.